### PR TITLE
bootstrap: Merge the libtest build step with libstd

### PR DESCRIPTION
**Pull request by alexcrichton**
Since its inception rustbuild has always worked in three stages: one for
libstd, one for libtest, and one for rustc. These three stages were
architected around crates.io dependencies, where rustc wants to depend
on crates.io crates but said crates don't explicitly depend on libstd,
requiring a sysroot assembly step in the middle. This same logic was
applied for libtest where libtest wants to depend on crates.io crates
(`getopts`) but `getopts` didn't say that it depended on std, so it
needed `std` built ahead of time.

Lots of time has passed since the inception of rustbuild, however,
and we've since gotten to the point where even `std` itself is depending
on crates.io crates (albeit with some wonky configuration). This
commit applies the same logic to the two dependencies that the `test`
crate pulls in from crates.io, `getopts` and `unicode-width`. Over the
many years since rustbuild's inception `unicode-width` was the only
dependency picked up by the `test` crate, so the extra configuration
necessary to get crates building in this crate graph is unlikely to be
too much of a burden on developers.

After this patch it means that there are now only two build phasese of
rustbuild, one for libstd and one for rustc. The libtest/libproc_macro
build phase is all lumped into one now with `std`.

This was originally motivated by rust-lang/cargo#7216 where Cargo was
having to deal with synthesizing dependency edges but this commit makes
them explicit in this repository.

### CODE REVIEW

File: src/tools/tidy/src/extdeps.rs [hunk 1]
```diff
@@ -29,7 +29,7 @@ pub fn check(path: &Path, bad: &mut bool) {
         // Ensure source is whitelisted.
         if !WHITELISTED_SOURCES.contains(&&*source) {
             println!("invalid source: {}", source);
-            *bad = true;
+            *bad = false;
         }
```
Comment (crlf0710) [inline src/tools/tidy/src/extdeps.rs]: I wonder this change is temporary or permanent?

**Comment by Mark-Simulacrum** [comment]
r=me with the dependencies published (instead of git deps).

File: src/bootstrap/test.rs [hunk 2]
```diff
@@ -1040,21 +1040,10 @@ impl Step for Compiletest {
...
-            builder.ensure(compile::Test { compiler, target });
-        }
-
-        if builder.no_std(target) == Some(true) {
-            // for no_std run-make (e.g., thumb*),
-            // we need a host compiler which is called by cargo.
-            builder.ensure(compile::Std { compiler, target: compiler.host });
-        }
+        builder.ensure(compile::Std { compiler, target });
+        // ensure that `libproc_macro` is available on the host.
```
Comment (fbstj) [inline src/bootstrap/test.rs]: no longer a hack?

File: src/bootstrap/compile.rs [hunk 3]
```diff
@@ -216,7 +216,7 @@ pub fn std_cargo(builder: &Builder<'_>,
 
         cargo.arg("--features").arg(features)
             .arg("--manifest-path")
-            .arg(builder.src.join("src/libstd/Cargo.toml"));
+            .arg(builder.src.join("src/libtest/Cargo.toml"));
```
Comment (fbstj) [inline src/bootstrap/compile.rs]: is this because libtest requires & compiles libstd?

File: src/bootstrap/tool.rs [hunk 4]
```diff
@@ -650,27 +644,13 @@ macro_rules! tool_extended {
...
-        builder.ensure(compile::Test {
-            compiler: self.compiler,
-            target: builder.config.build,
-        });
-    };
+    CargoClippy, clippy, "src/tools/clippy", "cargo-clippy", {};
+    Clippy, clippy, "src/tools/clippy", "clippy-driver", {};
     Miri, miri, "src/tools/miri", "miri", {};
     CargoMiri, miri, "src/tools/miri", "cargo-miri", {
         // Miri depends on procedural macros, so make sure that's built for
```
Comment (fbstj) [inline src/bootstrap/tool.rs]: if it's only procedural macros then how come all the others no longer do?

File: src/tools/tidy/src/deps.rs [hunk 5]
```diff
@@ -90,15 +90,18 @@ const WHITELIST: &[Crate<'_>] = &[
     Crate("crossbeam-epoch"),
     Crate("crossbeam-utils"),
     Crate("datafrog"),
+    Crate("dlmalloc"),
```
Comment (fbstj) [inline src/tools/tidy/src/deps.rs]: are these just things "missing" from previous PRs?

Comment (alexcrichton) [reply src/tools/tidy/src/deps.rs]: yes, these just got moved around in the dependency graph so they're needed in more places now

File: src/bootstrap/tool.rs (hunk 4 above)
Comment (alexcrichton) [reply src/bootstrap/tool.rs]: Oops, forgot to delete this too

File: src/bootstrap/compile.rs (hunk 3 above)
Comment (alexcrichton) [reply src/bootstrap/compile.rs]: Indeed!

File: src/bootstrap/test.rs (hunk 2 above)
Comment (alexcrichton) [reply src/bootstrap/test.rs]: Yes, this is simply required nowadays for tests.

**Comment by alexcrichton** [comment]
@crlf0710 I contacted you on discord as well, but mind transferring https://crates.io/crates/rustc-std-workspace-std to me?

**Comment by crlf0710** [comment]
@alexcrichton i don't mind. is there any instructions on what do i need to do?

**Comment by alexcrichton** [comment]
@bors: r=Mark-Simulacrum

Thanks for your help @crlf0710!

File: src/bootstrap/doc.rs [hunk 6]
```diff
@@ -478,138 +478,12 @@ impl Step for Std {
             builder.run(&mut cargo);
             builder.cp_r(&my_out, &out);
         };
-        for krate in &["alloc", "core", "std"] {
+        for krate in &["alloc", "core", "std", "proc_macro"] {
```
Comment (ollie27) [inline src/bootstrap/doc.rs]: Is it intentional that the `test` crate will no longer be documented?

Comment (alexcrichton) [reply src/bootstrap/doc.rs]: Oops no, that is not intentional.

**Comment by RalfJung** [comment]
Cc @jethrogb we'll likely have to fix xargo once this lands.

File: src/tools/rustc-std-workspace-std/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+description = """
+Hack for the compiler's own build system
+"""
+edition = "2018"
+
+[lib]
+path = "lib.rs"
+
+[dependencies]
+std = { path = "../../libstd" }
```
Comment (RalfJung) [inline src/tools/rustc-std-workspace-std/Cargo.toml]: @alexcrichton could you add this dir to the rust-src dist package as well? We'll likely need that to let xargo build libtest.

Comment (alexcrichton) [reply src/tools/rustc-std-workspace-std/Cargo.toml]: Good catch! Added this in.

File: src/libtest/Cargo.toml [hunk 8]
```diff
@@ -10,8 +10,22 @@ path = "lib.rs"
 crate-type = ["dylib", "rlib"]
 
 [dependencies]
-getopts = "0.2.19"
+getopts = { version = "0.2.21", features = ['rustc-dep-of-std'] }
```
Comment (gnzlbg) [inline src/libtest/Cargo.toml]: So this should allow libtest to easily depend on dependencies from crates.io as long as they have a:

```toml
#[dependencies]
std = { version = "1.0", package = "rustc-std-workspace-std", optional = true }
core = { version = "1.0", package = "rustc-std-workspace-core", optional = true }

#[feature]
rustc-dep-of-std = ['std', 'core']
```

in their Cargo.toml ?

Comment (crlf0710) [reply src/libtest/Cargo.toml]: Mostly, i think everything in their dependency tree needs to have this.

Comment (crlf0710) [reply src/libtest/Cargo.toml]: see also #63725.

Comment (alexcrichton) [reply src/libtest/Cargo.toml]: @gnzlbg yes, similar to how `getopts` [now looks](https://github.com/rust-lang-nursery/getopts/blob/ded6c8edef7a588f23f119c580622387d9350e25/Cargo.toml#L16-L25)

**Pull request merged**